compiler/next/include/chpl/AST/Variable.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_AST_VARIABLE_H
+#define CHPL_AST_VARIABLE_H
+
+#include "chpl/AST/Symbol.h"
+#include "chpl/AST/Location.h"
+
+namespace chpl {
+namespace ast {
+
+class Builder;
+
+/**
+  This class represents a variable. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      var a = 1;
+      ref b = a;
+      const c = 2;
+      const ref d = c;
+      param e = "hi";
+  \endrst
+
+  each of these is a VariableDecl that refers to a Variable Symbol.
+ */
+class Variable final : public Symbol {
+ friend class Builder;
+ friend class VariableDecl;
+
+ public:
+  enum Tag {
+    VAR,
+    CONST,
+    CONST_REF,
+    REF,
+    PARAM
+  };
+
+ private:
+  Tag tag_;
+  int8_t typeExprChildNum;
+  int8_t initExprChildNum;
+
+  Variable(ASTList children,
+           UniqueString name, Symbol::Visibility vis, Tag tag,
+           int8_t typeExprChildNum, int8_t initExprChildNum);
+  bool contentsMatchInner(const BaseAST* other) const override;
+
+ public:
+  ~Variable() override = default;
+  const Tag tag() const { return this->tag_; }
+  /**
+    Returns the type expression used in the variable's declaration, or nullptr
+    if there wasn't one.
+    */
+  const Expr* typeExpr() const {
+    if (typeExprChildNum >= 0) {
+      const BaseAST* ast = this->child(typeExprChildNum);
+      assert(ast->isExpr());
+      return (const Expr*)ast;
+    } else {
+      return nullptr;
+    }
+  }
+  /**
+    Returns the init expression used in the variable's declaration, or nullptr
+    if there wasn't one.
+    */
+  const Expr* initExpr() const {
+    if (initExprChildNum >= 0) {
+      const BaseAST* ast = this->child(initExprChildNum);
+      assert(ast->isExpr());
+      return (const Expr*)ast;
+    } else {
+      return nullptr;
+    }
+  }
+};
```

Comment:
I prefer `typeExpr` and `initExpr` to access fixed positions within `children_` vector because to me it would be more elegant.

If we do not want to store NULLs in `children_`, I propose storing instances of a class that specifically represents the absence of a child.

Reply:
That's an interesting idea and would certainly simplify this code. I'll have a look.

Reply:
@vasslitvinov - so, I thought a bit about this alternative, and I would like to stick with the offsets. Basically, makes the code implementing the AST node (Variable) more complex but makes the code working with the AST less complex.

I know of 4 ways to solve this problem:
 1. Store indices/offsets (like the code does in this PR)
 2. Store NULL for missing data e.g. if initExpr is not present
 3. Store some other placeholder AST node (e.g. MissingExpr) for missing data
 4. Wrap the initExpr/typeExpr in something that can optionally contain an expr (e.g. a Block `{ }` could mean no value but `{ expr; }` means the value is present.)

I don't like 2 and 3 because code visiting the AST now has to deal with this issue. E.g. when visiting children of a given node, NULL has to be handled. Even if a placeholder AST is used, this is still true. Vs with 1, this is not a problem. Additionally, there are other AST elements that need to represent optional fields. For example, a Function has any number of arguments, perhaps a return type, perhaps a where clause, and perhaps a body. This Function case would not work with 2 or 3 alone because there can be an arbitrary number of arguments before this optional data. (It would work with 4, where we would but the formals into some kind of Block-ish thing).

I don't like 4 because it adds more AST nodes - adding indirection in traversals and using more memory. Additionally it wouldn't be great to use Block for this because it'd be harder to know if something is a statement or not.

Regarding the potential problem with 1 that myVariable.child(0) might be a typeExpr and might be an initExpr - if the difference matters, the code should be using the `Variable` accessor methods typeExpr()/initExpr(). Otherwise it would be assuming something about the AST (that child(0) is a particular thing) that would not be true for a node type other than `Variable`.